Source/DataModels/Components/UI/ComponentVideo.cpp

```diff
@@ -0,0 +1,334 @@
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+	std::filesystem::path fs_path(video->GetAssetsPath());
+	std::string extension = fs_path.extension().string();
+	if (extension != ".avi")
+	{
+		canRotate = true;
+	}
+	OpenVideo(video->GetAssetsPath().c_str());
+}
+
+void ComponentVideo::Draw() const
+{
+	Program* program = App->GetModule<ModuleProgram>()->GetProgram(ProgramType::SPRITE);
+	if (program)
+	{
+		glEnable(GL_BLEND);
+		glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+
+		program->Activate();
+
+		ComponentTransform2D* transform = GetOwner()->GetComponentInternal<ComponentTransform2D>();
+
+		const float4x4& proj = App->GetModule<ModuleCamera>()->GetOrthoProjectionMatrix();
+		const float4x4& model = transform->GetGlobalScaledMatrix();
+		float4x4 view = float4x4::identity;
+
+		ComponentCanvas* canvas = transform->WhichCanvasContainsMe();
+		if (canvas)
+		{
+			canvas->RecalculateSizeAndScreenFactor();
+			float factor = canvas->GetScreenFactor();
+			view = view * float4x4::Scale(factor, factor, factor);
+		}
+
+		glUniformMatrix4fv(2, 1, GL_TRUE, (const float*) &view);
+		glUniformMatrix4fv(1, 1, GL_TRUE, (const float*) &model);
+		glUniformMatrix4fv(0, 1, GL_TRUE, (const float*) &proj);
+
+		glBindVertexArray(App->GetModule<ModuleUI>()->GetQuadVAO());
+
+		glActiveTexture(GL_TEXTURE0);
+		program->BindUniformFloat4("spriteColor", float4(1.0f, 1.0f, 1.0f, 1.0f));
+		program->BindUniformFloat("renderPercentage", 1.0f);
+		program->BindUniformInt("direction", 1);
+		
+		if (initialized)
+		{
+			program->BindUniformInt("hasDiffuse", 1);
+			glBindTexture(GL_TEXTURE_2D, frameTexture);
+			glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA8, frameWidth, frameHeight, 0, GL_RGBA, GL_UNSIGNED_BYTE, frameData);
+		}
+		else program->BindUniformInt("hasDiffuse", 0);
+
+		glDrawArrays(GL_TRIANGLES, 0, 6);
+
+		glBindTexture(GL_TEXTURE_2D, 0);
+		glBindVertexArray(0);
+		glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, 0);
+
+		glDisable(GL_BLEND);
+		program->Deactivate();
+	}
+}
+
+void ComponentVideo::InternalSave(Json& meta)
+{
+	meta["loop"] = loop;
+	meta["rotateVertical"] = rotateVertical;
+	meta["canBeRotate"] = canRotate;
+	meta["assetPathVideo"] = video->GetAssetsPath().c_str();
+}
+
+void ComponentVideo::InternalLoad(const Json& meta)
+{
+	loop = meta["loop"];
+	std::string path = meta["assetPathVideo"];
+	rotateVertical = meta["rotateVertical"];
+	canRotate = meta["canBeRotate"];
+	bool resourceExists = !path.empty() && App->GetModule<ModuleFileSystem>()->Exists(path.c_str());
+	if (resourceExists)
+	{
+		std::shared_ptr<ResourceVideo> resourceImage =
+			App->GetModule<ModuleResources>()->RequestResource<ResourceVideo>(path);
+		if (resourceImage)
+		{
+			video = resourceImage;
```

Comment:
Use move instead of copy, since it's more efficient.
```suggestion
			video = std::move(resourceImage);
```

Reply:
Done